File: 01. DataStructures/01. Linked List/1. Singly-Linked-List/InterviewBit Questions/12. Reverse LL-ii/solution_12.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode *Solution::reverseBetween(ListNode *A, int B, int C)
{
    if (B == C)
        return A;
    int cnt = 1;
    ListNode *start, *end, *pre, *head = A;
    while (cnt != B)
    {
        if (cnt + 1 == B)
            pre = head;
        head = head->next;
        cnt++;
    }
    start = head;
    while (cnt != C + 1)
    {
        head = head->next;
        cnt++;
    }
    end = head;
    ListNode *temp = NULL, *nex = start;
    head = start;
    while (start != end)
    {
        nex = start->next;
        start->next = temp;
        temp = start;
        start = nex;
    }
    if (B == 1)
        A = temp;
    else
        pre->next = temp;
    head->next = nex;
    return A;
}

File: 01. DataStructures/01. Linked List/1. Singly-Linked-List/InterviewBit Questions/12. Reverse LL-ii/solution_12.js
// REMEMBER 
// you need legit linked list for input to test this solution, which you don't have but you have checked it on leet code
// it's workin fine. just push it on github as it is. it's legit solution and it's working.


var reverseBetween = function(head, left, right) {
    
    let dummy = new ListNode(null, head);
    let preNode = dummy;
    let current = head;
      let count = 0;
      while(count < left - 1) {
          current = current.next;
          preNode = preNode.next;
          count++;
      }
      
      count = 0;
      let pre = null;
      while(count < right - left + 1) {
          const temp = current.next;
          current.next = pre;
          pre = current;
          current = temp;
          count++;
      }
      
      preNode.next.next = current;
      preNode.next = pre;
      return dummy.next;
  };
  
